standard/variables.md

```diff
@@ -1249,3 +1249,20 @@ A `new` expression that invokes a constructor obeys the same rules as a method i
 - Neither a `ref` local, nor a local of a `ref struct` type shall be in context at the point of a `yield return` statement or an `await` expression.
 - For a ref reassignment `e1 = ref e2`, the ref-safe-context of `e2` shall be at least as wide a context as the *ref-safe-context* of `e1`.
 - For a ref return statement `return ref e1`, the ref-safe-context of `e1` shall be the caller-context.
+
+### §Null-Tracking Null tracking
+
+For certain expressions denoting variables, fields, or properties, the null state (§Nullabilities-And-Null-States) is tracked between occurrences, based on assignments to them, tests performed on them, and the control flow between them. (This is similar to how definite assignment is tracked for variables.) Expressions that are tracked are, as follows:
+
+```ANTLR
+tracked_expression
+    : simple_name
+    | 'this'
+    | 'base'
+    | tracked_expression '.' identifier
+    ;
+```
+
+where *identifier* denotes a field or property.
+
+The null state for a tracked variable in unreachable code is “not null.” (This follows other decisions around unreachable code like considering all locals to be definitely assigned.)
```

Comment:
**[Disclaimer: I've yet to review this PR, Bill’s post just caught my eye]**

Yes, it feels incomplete, but maybe for different reasons to each of us ;-)

*tracked_expression* is not a production *per se* in the language – the tracked expressions are a subset of those defined by other productions, which are in the language, which also follow the grammar specified for *tracked_expression* (unless there are major grammar changes).

Is a variable reference such as `A<X>.B<Y>.c` tracked? If so it is not a *tracked_expression* by the above grammar.

*tracked_expression* is also left-recursive which we try to avoid, but that is simple to address.

I looked a bit further in this PR and line 1953 of `expression.md` refers to the static flow analysis as “the compiler’s” – I think it is probably “the language’s”… (the specification defines the language, a compiler (*et al*) implements the specification).

Reply:
**[A few days later…]**

It is **massively** incomplete, as it is in the source documents mentioned…

For example [latest source document mentioned](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/nullable-reference-types-specification.md) contains:

- “Describe null state transitions similar to definite assignment” as a placeholder, the whole flow analysis algorithm appears to be missing.
- “The null state of `E1 ? E2 : E3` is based on the null state of `E2` and `E3`…” – this is fine as far as it goes but as far as it goes does not cover the null state of a variable tested against `null` in `E1` when used in `E2` and `E3`.

So there does appear to a significant hole based on the currently considered source documents.

Now I don’t doubt that the group could devise an algorithm to fill this hole, but surely there is a spec for it?